executor/analyze.go

```diff
@@ -64,74 +63,179 @@ func (e *AnalyzeExec) Close() error {
 
 // Next implements the Executor Next interface.
 func (e *AnalyzeExec) Next() (*Row, error) {
-	for _, src := range e.Srcs {
-		ae := src.(*AnalyzeExec)
-		var count int64 = -1
-		var sampleRows []*ast.Row
-		var colNDVs []int64
-		if len(ae.colIDs) != 0 {
-			rs := &recordSet{executor: ae.Srcs[len(ae.Srcs)-1]}
-			var err error
-			count, sampleRows, colNDVs, err = CollectSamplesAndEstimateNDVs(rs, len(ae.colIDs))
-			if err != nil {
-				return nil, errors.Trace(err)
+	var tasks []*analyzeTask
+	for i, src := range e.Srcs {
+		switch src.(type) {
+		case *XSelectTableExec:
+			if i < e.pkCount {
+				tasks = append(tasks, e.buildAnalyzePKTask(src.(*XSelectTableExec)))
+			} else {
+				t, err := e.buildAnalyzeColumnsTask(src.(*XSelectTableExec))
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				tasks = append(tasks, t...)
 			}
+		case *XSelectIndexExec:
+			tasks = append(tasks, e.buildAnalyzeIndexTask(src.(*XSelectIndexExec)))
 		}
-		columnSamples := rowsToColumnSamples(sampleRows)
-		var pkRS ast.RecordSet
-		if ae.pkID != 0 {
-			offset := len(ae.Srcs) - 1
-			if len(ae.colIDs) != 0 {
-				offset--
-			}
-			pkRS = &recordSet{executor: ae.Srcs[offset]}
+	}
+	concurrency, err := getBuildStatsConcurrency(e.ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	taskCh := make(chan *analyzeTask, len(tasks))
+	resultCh := make(chan *analyzeResult, len(tasks))
+	for i := 0; i < concurrency; i++ {
+		go analyzeWorker(taskCh, resultCh)
+	}
+	for _, task := range tasks {
+		taskCh <- task
+	}
+	close(taskCh)
+	for i := 0; i < len(tasks); i++ {
+		result := <-resultCh
+		if result.err != nil {
+			return nil, errors.Trace(err)
+		}
+		err = result.hist.SaveToStorage(e.ctx, result.tblInfo.ID, result.count, result.isIndex)
+		if err != nil {
+			return nil, errors.Trace(err)
 		}
-		idxRS := make([]ast.RecordSet, 0, len(ae.idxIDs))
-		for i := range ae.idxIDs {
-			idxRS = append(idxRS, &recordSet{executor: ae.Srcs[i]})
+		// The sorted bucket is more accurate, we replace the sampled column histogram with index histogram if the
```

Comment:
Indeed, we can remove this logic. @winoros is designing a algo : Given some expressions and some histgorams, calcute the selectitvy. This algorithm will choose index histogram as possible as it can,